speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp

```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
+
+    // Write a file with dummy initialization data, save timestamp 0 as many times as needed, and shut down
     FileDataStorage fileDataStorage;
-    fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
-    fileDataStorage.Save(nullptr, nullptr, 0);
+    std::experimental::filesystem::path actualPath = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+
+    for (int i = 0; i < p_numberOfTicks; i++)
+    {
+        fileDataStorage.Save(nullptr, nullptr, 0);
+        expected << "0\n";
+    }
     fileDataStorage.Shutdown();
+    expected << "END";
 
+    // Check if the file is truly written to the Windows temporary directory
     std::experimental::filesystem::path path = std::experimental::filesystem::temp_directory_path();
-    path.append(TEST_FILE_PATH);
-    std::string stringpath = path.string();
+    path.append(TEST_FILE_DIR);
+    path.append(TEST_FILE_NAME);
+    std::string expectedPath = path.string();
+    ASSERT_STREQ(expectedPath.c_str(), actualPath.string().c_str());
+
+    // Read the written file
+    std::ifstream reader(actualPath.c_str());
+    std::string fileContents((std::istreambuf_iterator<char>(reader)), std::istreambuf_iterator<char>());
+    reader.close();
+
+    // Check contents
+    ASSERT_STREQ(fileContents.c_str(), expected.str().c_str());
+}
+
+/// @brief Test the data storage system's Save function over 3 timesteps with certain modules enabled.
+/// Data used will be randomly generated.
+/// No decisions will be tested.
+/// @param p_storeEnvironment Whether to save environment data
+/// @param p_storeCar Whether to save car data
+/// @param p_storeControls Whether to save player control data
+/// @param p_storeDecisions Whether to save intervention data
+/// @param p_storeMeta Whether to save metadata
+void TestDataStorageSave(bool p_storeEnvironment, bool p_storeCar, bool p_storeControls, bool p_storeDecisions, bool p_storeMeta)
+{
+    Random random;
+    tDataToStore params = {
+        p_storeEnvironment,
+        p_storeCar,
+        p_storeControls,
+        p_storeDecisions,
+        p_storeMeta};
+    FileDataStorage fileDataStorage;
+
+    // Create a string to save all intended random data to
+    std::stringstream expected;
+
+    // Initialise buffer file
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+    if (p_storeEnvironment)
+    {
+        expected << "GameState\n";
+    }
+    if (p_storeControls)
+    {
+        expected << "UserInput\n";
+    }
+
+    for (int i = 0; i < 3; i++)
+    {
+        // Save with random data
+        GET_RANDOM_TICKCOUNT;
+        GET_RANDOM_CAR;
+        GET_RANDOM_SITUATION;
+        fileDataStorage.Save(&car, &situation, tickCount);
+
+        // Define our expectations
+        expected << std::to_string(tickCount) << "\n";
```

Comment:
Change all '\n' characters with std::endl as enters are different on windows and linux